# Lab 01

Welcome to the first lab of Data Science Foundations! This lab is meant to help you familiarize yourself with Jupyter Notebook, review Python and `numpy`, and `Pandas`.


---
## Part 1: Jupyter Tips


### Viewing Documentation

To output the documentation for a function, use the `help` function.

In [313]:
help(print)

Help on built-in function print in module builtins:

print(...)
    print(value, ..., sep=' ', end='\n', file=sys.stdout, flush=False)
    
    Prints the values to a stream, or to sys.stdout by default.
    Optional keyword arguments:
    file:  a file-like object (stream); defaults to the current sys.stdout.
    sep:   string inserted between values, default a space.
    end:   string appended after the last value, default a newline.
    flush: whether to forcibly flush the stream.



### Importing Libraries and Magic Commands

In Data Science Foundations, we will be using common Python libraries to help us process data. By convention, we import all libraries at the very top of the notebook. There are also a set of standard aliases that are used to shorten the library names. Below are some of the libraries that you may encounter throughout the course, along with their respective aliases.

In [314]:
import pandas as pd
import IPython.display
import numpy as np
import matplotlib.pyplot as plt
plt.style.use('fivethirtyeight')
%matplotlib inline

### Keyboard Shortcuts

Even if you are familiar with Jupyter, we strongly encourage you to become proficient with keyboard shortcuts (this will save you time in the future). To learn about keyboard shortcuts, go to **Help --> Keyboard Shortcuts** in the menu above.

Here are a few that we like:
1. `Ctrl` + `Return` (or `Cmd` + `Return` on Mac): *Evaluate the current cell*
1. `Shift` + `Return`: *Evaluate the current cell and move to the next*
1. `ESC` : *command mode* (may need to press before using any of the commands below)
1. `a` : *create a cell above*
1. `b` : *create a cell below*
1. `dd` : *delete a cell*
1. `z` : *undo the last cell operation*
1. `m` : *convert a cell to markdown*
1. `y` : *convert a cell to code*

---
## Part 2: Prerequisites

It's time to answer some review questions. Each question has a response cell directly below it. Most response cells are followed by a test cell that runs automated tests to check your work. Please don't delete questions, response cells, or test cells. You won't get credit for your work if you do.

If you have extra content in a response cell, such as an example call to a function you're implementing, that's fine. Also, feel free to add cells between the question cells and test cells (or the next cell, for questions without test cases). Any extra cells you add will be considered part of your submission. Finally, when you finish an assignment, make sure to "restart and run all cells" to ensure everything works properly.

Note that for labs, ontime submissions that pass all the test cases will receive full credit. However for homeworks, test cells don't always confirm that your response is correct. They are meant to give you some useful feedback, but it's your responsibility to ensure your response answers the question correctly. There may be other tests that we run when scoring your notebooks. We **strongly recommend** that you check your solutions yourself rather than just relying on the test cells.

### Python

Python is the main programming language we'll use in the course. We expect that you've taken Python programming courses, or an equivalent class, so we will not be covering general Python syntax. If any of the following exercises are challenging (or if you would like to refresh your Python knowledge), please review one or more of the following materials.

- **[Python Tutorial](https://docs.python.org/3.8/tutorial/)**: Introduction to Python from the creators of Python.
- **[Composing Programs Chapter 1](http://composingprograms.com/pages/11-getting-started.html)**: This is more of a introduction to programming with Python.
- **[Advanced Crash Course](http://cs231n.github.io/python-numpy-tutorial/)**: A fast crash course which assumes some programming background.

### NumPy

NumPy is the numerical computing module. Here's a quick recap of NumPy. For more review, read the following materials.

- **[NumPy Quick Start Tutorial](https://numpy.org/doc/stable/user/quickstart.html)**
- **[DS100 NumPy Review](http://ds100.org/fa17/assets/notebooks/numpy/Numpy_Review.html)**
- **[Stanford CS231n NumPy Tutorial](http://cs231n.github.io/python-numpy-tutorial/#numpy)**
- **[The Data 8 Textbook Chapter on NumPy](https://www.inferentialthinking.com/chapters/05/1/Arrays)**

### Question 1

The core of NumPy is the array. Like Python lists, arrays store data; however, they store data in a more efficient manner. In many cases, this allows for faster computation and data manipulation.

Use `np.array` to create an array. It takes a sequence, such as a list or range.

Below, create an array `arr` containing the values 1, 2, 3, 4, and 5 (in that order).

<!--
BEGIN QUESTION
name: q1
points: 1
-->

In [315]:
arr = np.array([1, 2, 3, 4, 5])

In [316]:
assert (arr == np.array([1, 2, 3, 4, 5])).all()
assert type(arr) is np.ndarray

In addition to values in the array, we can access attributes such as shape and data type. A full list of attributes can be found [here](https://docs.scipy.org/doc/numpy-1.15.0/reference/arrays.ndarray.html#array-attributes).

In [317]:
arr[3]

4

In [318]:
arr[2:4]

array([3, 4])

In [319]:
arr.shape

(5,)

In [320]:
arr.dtype

dtype('int64')

Arrays, unlike Python lists, cannot store items of different data types.

In [321]:
# A regular Python list can store items of different data types
[1, '3']

[1, '3']

In [322]:
# Arrays will convert everything to the same data type
np.array([1, '3'])

array(['1', '3'], dtype='<U21')

In [323]:
# Another example of array type conversion
np.array([5, 8.3])

array([5. , 8.3])

Arrays are also useful in performing *vectorized operations*. Given two or more arrays of equal length, arithmetic will perform element-wise computations across the arrays.

For example, observe the following:

In [324]:
# Python list addition will concatenate the two lists
[1, 2, 3] + [4, 5, 6]

[1, 2, 3, 4, 5, 6]

In [325]:
# NumPy array addition will add them element-wise
np.array([1, 2, 3]) + np.array([4, 5, 6])

array([5, 7, 9])

### Question 2

#### Question 2a

Write a function `summation` that evaluates the following summation for $n \geq 1$:

$$\sum_{i=1}^{n} i^3 + 3 i^2$$

**Note**: You should not use `for` loops in your solution. Check the [NumPy documentation](https://numpy.org/doc/1.19/reference/index.html). If you're stuck, try a search engine! Searching the web for examples of how to use modules is very common in data science.

<!--
BEGIN QUESTION
name: q2a
points: 1
-->

In [326]:
def summation(n):
    """Compute the summation i^3 + 3 * i^2 for 1 <= i <= n."""
    arr = np.array(range(1, n+1))
    return sum(arr ** 3 + 3 * arr ** 2)

In [327]:
assert summation(1) == 4
assert summation(2) == 24
assert summation(100) == 26517550

#### Question 2b

Write a function `elementwise_array_sum` that computes the square of each value in `list_1`, the cube of each value in `list_2`, then returns a list containing the element-wise sum of these results. Assume that `list_1` and `list_2` have the same number of elements, do not use for loops.

The input parameters will both be **python lists**, so you may need to convert the lists into arrays before performing your operations. The output should be a **numpy array.**



<!--
BEGIN QUESTION
name: q2b
points: 1
-->

In [328]:
def elementwise_array_sum(list_1, list_2):
    """Compute x^2 + y^3 for each x, y in list_1, list_2.

    Assume list_1 and list_2 have the same length.

    Return a NumPy array.
    """
    assert len(list_1) == len(list_2), "both args must have the same number of elements"
    arr1 = np.array(list_1)
    arr2 = np.array(list_2)
    arr1 = arr1**2
    arr2 = arr2**3
    return arr1+arr2

In [329]:
assert elementwise_array_sum([1], [1]) == np.array([2])
assert elementwise_array_sum([-1], [1]) == np.array([2])
assert elementwise_array_sum([1], [-1]) == np.array([0])
assert (elementwise_array_sum([1, 2, 3], [1, 2, 3]) == np.array([ 2, 12, 36])).all()
assert (elementwise_array_sum([1, 5, 2], [3, 6, 6]) == np.array([ 28, 241, 220])).all()
assert type(elementwise_array_sum([], [])) is np.ndarray

You might have been told that Python is slow, but array arithmetic is carried out very fast, even for large arrays. Below is an implementation of the above code that does not use NumPy arrays.

In [330]:
def elementwise_list_sum(list_1, list_2):
    """Compute x^2 + y^3 for each x, y in list_1, list_2.

    Assume list_1 and list_2 have the same length.
    """

    return [x ** 2 + y ** 3 for x, y in zip(list_1, list_2)]

For ten numbers, `elementwise_list_sum` and `elementwise_array_sum` both take a similar amount of time.

In [331]:
sample_list_1 = list(range(10))
sample_array_1 = np.arange(10)

In [332]:
%%time
elementwise_list_sum(sample_list_1, sample_list_1)

CPU times: user 11 µs, sys: 1 µs, total: 12 µs
Wall time: 13.8 µs


[0, 2, 12, 36, 80, 150, 252, 392, 576, 810]

In [333]:
%%time
elementwise_array_sum(sample_array_1, sample_array_1)

CPU times: user 34 µs, sys: 18 µs, total: 52 µs
Wall time: 47.9 µs


array([  0,   2,  12,  36,  80, 150, 252, 392, 576, 810])

The time difference seems negligible for a list/array of size 10; depending on your setup, you may even observe that `elementwise_list_sum` executes faster than `elementwise_array_sum`! However, we will commonly be working with much larger datasets:

In [334]:
sample_list_2 = list(range(100000))
sample_array_2 = np.arange(100000)

In [335]:
%%time
elementwise_list_sum(sample_list_2, sample_list_2)
; # The semicolon hides the output

CPU times: user 57.9 ms, sys: 1.77 ms, total: 59.7 ms
Wall time: 58.5 ms


'# The semicolon hides the output'

In [336]:
%%time
elementwise_array_sum(sample_array_2, sample_array_2)
;

CPU times: user 1.47 ms, sys: 1.02 ms, total: 2.49 ms
Wall time: 1.2 ms


''

With the larger dataset, we see that using NumPy results in code that executes over 50 times faster! Throughout this course (and in the real world), you will find that writing efficient code will be important; arrays and vectorized operations are the most common way of making Python programs run quickly.

#### Question 2d

Given the array `random_arr`, assign `valid_values` to an array containing all values $x$ such that $2x^4 > 1$.

**Note**: You should not use `for` loops in your solution. Instead, look at `numpy`'s documentation on [Boolean Indexing](https://numpy.org/devdocs/reference/arrays.indexing.html).

<!--
BEGIN QUESTION
name: q2d
points: 1
-->

In [337]:
np.random.seed(42)
random_arr = np.random.rand(60)
boolean_index = 2 * random_arr ** 4 > 1
valid_values = random_arr[boolean_index]

In [338]:
assert len(valid_values) == 10
assert np.allclose(valid_values, [0.95071431, 0.86617615, 0.96990985, 0.94888554, 0.96563203, 0.9093204 , 0.96958463, 0.93949894, 0.89482735, 0.92187424])

## Lab 2: Pandas Overview

---
[Pandas](https://pandas.pydata.org/) is one of the most widely used Python libraries in data science. In this lab, you will review commonly used data wrangling operations/tools in Pandas. We aim to give you familiarity with:

* Creating DataFrames
* Slicing DataFrames (i.e. selecting rows and columns)
* Filtering data (using boolean arrays and groupby.filter)
* Aggregating (using groupby.agg)

In this lab you are going to use several pandas methods. Reminder from lecture that you may press `shift+tab` on method parameters to see the documentation for that method. For example, if you were using the `drop` method in pandas, you could press shift+tab to see what `drop` is expecting.

This lab expects that you have watched the pandas lectures. If you have not, this lab will probably take a very long time.

**Note**: The Pandas interface is notoriously confusing for beginners, and the documentation is not consistently great. Throughout the semester, you will have to search through Pandas documentation and experiment, but remember it is part of the learning experience and will help shape you as a data scientist!

## Creating DataFrames & Basic Manipulations

Recall that a [DataFrame](https://pandas.pydata.org/pandas-docs/stable/user_guide/dsintro.html#dataframe) is a table in which each column has a specific data type; there is an index over the columns (typically string labels) and an index over the rows (typically ordinal numbers).

Usually you'll create DataFrames by using a function like `pd.read_csv`. However, in this section, we'll discuss how to create them from scratch.

The [documentation](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.html) for the pandas `DataFrame` class provides several constructors for the DataFrame class.

**Syntax 1:** You can create a DataFrame by specifying the columns and values using a dictionary as shown below.

The keys of the dictionary are the column names, and the values of the dictionary are lists containing the row entries.

In [339]:
fruit_info = pd.DataFrame(
    data = {'fruit': ['apple', 'orange', 'banana', 'raspberry'],
          'color': ['red', 'orange', 'yellow', 'pink'],
          'price': [1.0, 0.75, 0.35, 0.05]
          })
fruit_info

fruit   color  price
0      apple     red   1.00
1     orange  orange   0.75
2     banana  yellow   0.35
3  raspberry    pink   0.05

**Syntax 2:** You can also define a DataFrame by specifying the rows as shown below.

Each row corresponds to a distinct tuple, and the columns are specified separately.

In [340]:
fruit_info2 = pd.DataFrame(
    [("red", "apple", 1.0), ("orange", "orange", 0.75), ("yellow", "banana", 0.35),
     ("pink", "raspberry", 0.05)],
    columns = ["color", "fruit", "price"])
fruit_info2

color      fruit  price
0     red      apple   1.00
1  orange     orange   0.75
2  yellow     banana   0.35
3    pink  raspberry   0.05

You can obtain the dimensions of a DataFrame by using the shape attribute `DataFrame.shape`.

In [341]:
fruit_info.shape

(4, 3)

You can also convert the entire DataFrame into a two-dimensional NumPy array.

In [342]:
fruit_info.values

array([['apple', 'red', 1.0],
       ['orange', 'orange', 0.75],
       ['banana', 'yellow', 0.35],
       ['raspberry', 'pink', 0.05]], dtype=object)

There are other constructors but we do not discuss them here.

### REVIEW: Selecting Rows and Columns in Pandas

As you've seen in lecture and discussion, there are two verbose operators in Python for selecting rows: `loc` and `iloc`. Let's review them briefly.

#### Approach 1: `loc`

The first of the two verbose operators is `loc`, which takes two arguments. The first is one or more row **labels**, the second is one or more column **labels**.

The desired rows or columns can be provided individually, in slice notation, or as a list. Some examples are given below.

Note that **slicing in `loc` is inclusive** on the provided labels.

In [343]:
#get rows 0 through 2 and columns fruit through price
fruit_info.loc[0:2, 'fruit':'price']

fruit   color  price
0   apple     red   1.00
1  orange  orange   0.75
2  banana  yellow   0.35

In [344]:
# get rows 0 through 2 and columns fruit and price.
# Note the difference in notation and result from the previous example.
fruit_info.loc[0:2, ['fruit', 'price']]

fruit  price
0   apple   1.00
1  orange   0.75
2  banana   0.35

In [345]:
# get rows 0 and 2 and columns fruit and price.
fruit_info.loc[[0, 2], ['fruit', 'price']]

fruit  price
0   apple   1.00
2  banana   0.35

In [346]:
# get rows 0 and 2 and column fruit
fruit_info.loc[[0, 2], ['fruit']]

fruit
0   apple
2  banana

Note that if we request a single column but don't enclose it in a list, the return type of the `loc` operator is a `Series` rather than a DataFrame.

In [347]:
# get rows 0 and 2 and column fruit, returning the result as a Series
fruit_info.loc[[0, 2], 'fruit']

0     apple
2    banana
Name: fruit, dtype: object

If we provide only one argument to `loc`, it uses the provided argument to select rows, and returns all columns.

In [348]:
fruit_info.loc[0:1]

fruit   color  price
0   apple     red   1.00
1  orange  orange   0.75

#### Approach 2: `iloc`

`iloc` is very similar to `loc` except that its arguments are row numbers and column numbers, rather than row labels and labels names. A usueful mnemonic is that the `i` stands for "integer".

In addition, **slicing for `iloc` is exclusive** on the provided integer indices. Some examples are given below:

In [349]:
# get rows 0 through 3 (exclusive) and columns 0 through 2 (exclusive)
fruit_info.iloc[0:3, 0:3]

fruit   color  price
0   apple     red   1.00
1  orange  orange   0.75
2  banana  yellow   0.35

In [350]:
# get rows 0 through 3 (exclusive) and columns 0 and 2.
fruit_info.iloc[0:3, [0, 2]]

fruit  price
0   apple   1.00
1  orange   0.75
2  banana   0.35

In [351]:
# get rows 0 and 2 and columns 0 and 2.
fruit_info.iloc[[0, 2], [0, 2]]

fruit  price
0   apple   1.00
2  banana   0.35

In [352]:
#get rows 0 and 2 and column fruit
fruit_info.iloc[[0, 2], [0]]

fruit
0   apple
2  banana

In [353]:
# get rows 0 and 2 and column fruit
fruit_info.iloc[[0, 2], 0]

0     apple
2    banana
Name: fruit, dtype: object

Note that in these loc and iloc examples above, the row **label** and row **number** were always the same.

Let's see an example where they are different. If we sort our fruits by price, we get:

In [354]:
fruit_info_sorted = fruit_info.sort_values("price")
fruit_info_sorted

fruit   color  price
3  raspberry    pink   0.05
2     banana  yellow   0.35
1     orange  orange   0.75
0      apple     red   1.00

Observe that the row number 0 now has index 3, row number 1 now has index 2, etc. These indices are the arbitrary numerical index generated when we created the DataFrame. For example, banana was originally in row 2, and so it has row label 2.

If we request the rows in positions 0 and 2 using `iloc`, we're indexing using the row NUMBERS, not labels.

In [355]:
fruit_info_sorted.iloc[[0, 2], 0]

3    raspberry
1       orange
Name: fruit, dtype: object

Lastly, similar to with `loc`, the second argument to `iloc` is optional. That is, if you provide only one argument to `iloc`, it treats the argument you provide as a set of desired row numbers, not column numbers.

In [356]:
fruit_info.iloc[[0, 2]]

fruit   color  price
0   apple     red   1.00
2  banana  yellow   0.35

#### Approach 3: `[]` Notation for Accessing Rows and Columns

Pandas also supports a bare `[]` operator. It's similar to `loc` in that it lets you access rows and columns by their name.

However, unlike `loc`, which takes row names and also optionally column names, `[]` is more flexible. If you provde it only row names, it'll give you rows (same behavior as `loc`), and if you provide it with only column names, it'll give you columns (whereas `loc` will crash).

Some examples:

In [357]:
fruit_info[0:2]

fruit   color  price
0   apple     red   1.00
1  orange  orange   0.75

In [358]:
# Here we're providing a list of fruits as single argument to []
fruit_info[["fruit", "color", "price"]]

fruit   color  price
0      apple     red   1.00
1     orange  orange   0.75
2     banana  yellow   0.35
3  raspberry    pink   0.05

Note that slicing notation is not supported for columns if you use `[]` notation. Use `loc` instead.

In [359]:
# uncomment and this code crashes
#fruit_info["fruit":"price"]

# uncomment and this works fine
#fruit_info.loc[:, "fruit":"price"]

`[]` and `loc` are quite similar. For example, the following two pieces of code are functionally equivalent for selecting the fruit and price columns.

1. `fruit_info[["fruit", "price"]]`
2. `fruit_info.loc[:, ["fruit", "price"]]`.

Because it yields more concise code, you'll find that our code and your code both tend to feature `[]`. However, there are some subtle pitfalls of using `[]`. If you're ever having performance issues, weird behavior, or you see a `SettingWithCopyWarning` in pandas, switch from `[]` to `loc` and this may help.

To avoid getting too bogged down in indexing syntax, we'll avoid a more thorough discussion of `[]` and `loc`. We may return to this at a later point in the course.

For more on `[]` vs `loc`, you may optionally try reading:
1. https://stackoverflow.com/questions/48409128/what-is-the-difference-between-using-loc-and-using-just-square-brackets-to-filte
2. https://stackoverflow.com/questions/38886080/python-pandas-series-why-use-loc/65875826#65875826
3. https://stackoverflow.com/questions/20625582/how-to-deal-with-settingwithcopywarning-in-pandas/53954986#53954986

Now that we've reviewed basic indexing, let's discuss how we can modify dataframes. We'll do this via a series of exercises.

### Question 1

For a DataFrame `d`, you can add a column with `d['new column name'] = ...` and assign a list or array of values to the column. Add a column of integers containing 1, 2, 3, and 4 called `rank` to the `fruit_info` table which expresses your personal preference about the taste ordering for each fruit (1 is tastiest; 4 is least tasty).

<!--
BEGIN QUESTION
name: q1a
-->

In [360]:
fruit_info['rank'] = [2, 1, 4, 3]
fruit_info

fruit   color  price  rank
0      apple     red   1.00     2
1     orange  orange   0.75     1
2     banana  yellow   0.35     4
3  raspberry    pink   0.05     3

In [361]:
assert fruit_info["rank"].dtype == np.dtype('int64')
assert sorted(fruit_info["rank"].dropna()) == [1, 2, 3, 4]

### Question 2

Use the `.drop()` method to [drop](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.drop.html) the `rank` column you created. Make sure to use the `axis` parameter correctly. Note that `drop` does not change a table, but instead returns a new table with fewer columns or rows unless you set the optional `inplace` parameter.

*Hint*: Look through the [documentation](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.drop.html) to see how you can drop multiple columns of a Pandas DataFrame at once using a list of column names.

<!--
BEGIN QUESTION
name: q2
-->

In [362]:
fruit_info_original = fruit_info.drop(labels = ['rank'], axis = 1)
fruit_info_original

fruit   color  price
0      apple     red   1.00
1     orange  orange   0.75
2     banana  yellow   0.35
3  raspberry    pink   0.05

In [363]:
assert fruit_info_original.shape == (4, 3)
assert fruit_info.shape == (4, 4)
assert (fruit_info_original.columns == np.array(['fruit', 'color', 'price'], dtype='object')).all()

### Question 3

Use the `.rename()` method to [rename](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.rename.html) the columns of `fruit_info_original` so they begin with capital letters. Set this new DataFrame to `fruit_info_caps`. For an example of how to use rename, see the linked documentation above.
<!--
BEGIN QUESTION
name: q3
-->

In [364]:
fruit_info_caps = fruit_info_original.rename(columns = {'fruit':'Fruit', 'color':'Color', 'price':'Price'})
fruit_info_caps

Fruit   Color  Price
0      apple     red   1.00
1     orange  orange   0.75
2     banana  yellow   0.35
3  raspberry    pink   0.05

In [365]:
assert (fruit_info_caps.columns == np.array(['Fruit', 'Color', 'Price'], dtype='object')).all()
assert (fruit_info.columns == np.array(['fruit', 'color', 'price', 'rank'], dtype='object')).all()

### Babynames Dataset
For the new few questions of this lab, let's move on to a real world dataset. We'll be using the babynames dataset from Lecture 1. The babynames dataset contains a record of the given names of babies born in the United States each year.

First let's run the following cells to build the DataFrame `baby_names`.
The cells below download the data from the web and extract the data into a DataFrame. There should be a total of 6215834 records.

### `fetch_and_cache` Helper

The following function downloads and caches data in the `data/` directory and returns the `Path` to the downloaded file. The cell below the function describes how it works. You are not expected to understand this code, but you may find it useful as a reference as a practitioner of data science after the course.

In [366]:
import requests
from pathlib import Path

def fetch_and_cache(data_url, file, data_dir="data", force=False):
    """
    Download and cache a url and return the file object.

    data_url: the web address to download
    file: the file in which to save the results.
    data_dir: (default="data") the location to save the data
    force: if true the file is always re-downloaded

    return: The pathlib.Path to the file.
    """
    data_dir = Path(data_dir)
    data_dir.mkdir(exist_ok=True)
    file_path = data_dir/Path(file)
    if force and file_path.exists():
        file_path.unlink()
    if force or not file_path.exists():
        print('Downloading...', end=' ')
        resp = requests.get(data_url)
        with file_path.open('wb') as f:
            f.write(resp.content)
        print('Done!')
    else:
        import time
        created = time.ctime(file_path.stat().st_ctime)
        print("Using cached version downloaded at", created)
    return file_path

In Python, a `Path` object represents the filesystem paths to files (and other resources). The `pathlib` module is effective for writing code that works on different operating systems and filesystems.

To check if a file exists at a path, use `.exists()`. To create a directory for a path, use `.mkdir()`. To remove a file that might be a [symbolic link](https://en.wikipedia.org/wiki/Symbolic_link), use `.unlink()`.

This function creates a path to a directory that will contain data files. It ensures that the directory exists (which is required to write files in that directory), then proceeds to download the file based on its URL.

The benefit of this function is that not only can you force when you want a new file to be downloaded using the `force` parameter, but in cases when you don't need the file to be re-downloaded, you can use the cached version and save download time.

Below we use `fetch_and_cache` to download the `namesbystate.zip` zip file, which is a compressed directory of CSV files.

**This might take a little while! Consider stretching.**

In [367]:
data_url = 'https://www.ssa.gov/oact/babynames/state/namesbystate.zip'
namesbystate_path = fetch_and_cache(data_url, 'namesbystate.zip')

Using cached version downloaded at Tue Mar 21 14:46:09 2023


The following cell builds the final full `baby_names` DataFrame. It first builds one DataFrame per state, because that's how the data are stored in the zip file. Here is documentation for [pd.concat](https://pandas.pydata.org/pandas-docs/version/1.2/reference/api/pandas.concat.html) if you want to know more about its functionality. As before, you are not expected to understand this code.

In [368]:
import zipfile
zf = zipfile.ZipFile(namesbystate_path, 'r')

column_labels = ['State', 'Sex', 'Year', 'Name', 'Count']

def load_dataframe_from_zip(zf, f):
    with zf.open(f) as fh:
        return pd.read_csv(fh, header=None, names=column_labels)

states = [
    load_dataframe_from_zip(zf, f)
    for f in sorted(zf.filelist, key=lambda x:x.filename)
    if f.filename.endswith('.TXT')
]

baby_names = states[0]
for state_df in states[1:]:
    baby_names = pd.concat([baby_names, state_df])
baby_names = baby_names.reset_index().iloc[:, 1:]

In [369]:
len(baby_names)

6215834

In [370]:
baby_names.head()

State Sex  Year      Name  Count
0    AK   F  1910      Mary     14
1    AK   F  1910     Annie     12
2    AK   F  1910      Anna     10
3    AK   F  1910  Margaret      8
4    AK   F  1910     Helen      7

### Question 4

Use `.loc` to select `Name` and `Year` **in that order** from the `baby_names` table.

<!--
BEGIN QUESTION
name: q4
-->

In [371]:
name_and_year = baby_names.loc[:, ['Name', 'Year']]
name_and_year[:5]

Name  Year
0      Mary  1910
1     Annie  1910
2      Anna  1910
3  Margaret  1910
4     Helen  1910

In [372]:
assert name_and_year.shape == (6215834, 2)
assert name_and_year.loc[0, "Name"] == 'Mary'
assert name_and_year.loc[0, "Year"] == 1910

## Filtering Data

### Review: Filtering with boolean arrays

Filtering is the process of removing unwanted material.  In your quest for cleaner data, you will undoubtedly filter your data at some point: whether it be for clearing up cases with missing values, for culling out fishy outliers, or for analyzing subgroups of your data set.  Example usage looks like `df[df['column name'] < 5]`.

For your reference, some commonly used comparison operators are given below.

Symbol | Usage      | Meaning
------ | ---------- | -------------------------------------
==   | a == b   | Does a equal b?
<=   | a <= b   | Is a less than or equal to b?
&gt;=   | a >= b   | Is a greater than or equal to b?
<    | a < b    | Is a less than b?
&#62;    | a &#62; b    | Is a greater than b?
~    | ~p       | Returns negation of p
&#124; | p &#124; q | p OR q
&    | p & q    | p AND q
^  | p ^ q | p XOR q (exclusive or)

In the following we construct the DataFrame containing only names registered in California

In [373]:
ca = baby_names[baby_names['State'] == 'CA']
ca.head(5)

State Sex  Year      Name  Count
390635    CA   F  1910      Mary    295
390636    CA   F  1910     Helen    239
390637    CA   F  1910   Dorothy    220
390638    CA   F  1910  Margaret    163
390639    CA   F  1910   Frances    134

### Question 5
Using a boolean array, select the names in Year 2000 (from `baby_names`) that have larger than 3000 counts. Keep all columns from the original `baby_names` DataFrame.

Note: Note that compound expressions have to be grouped with parentheses. That is, any time you use `p & q` to filter the DataFrame, make sure to use `df[(df[p]) & (df[q])]` or `df.loc[(df[p]) & (df[q])]`.

You may use either `[]` or `loc`. Both will achieve the same result. For more on `[]` vs. `loc` see the stack overflow links from the intro portion of this lab.

<!--
BEGIN QUESTION
name: q5
-->

In [374]:
result = baby_names[(baby_names['Count'] > 3000) & (baby_names['Year'] == 2000)]
result.head()

State Sex  Year     Name  Count
725638    CA   M  2000   Daniel   4342
725639    CA   M  2000  Anthony   3839
725640    CA   M  2000     Jose   3804
725641    CA   M  2000   Andrew   3600
725642    CA   M  2000  Michael   3572

In [375]:
assert len(result) == 11
assert result["Count"].sum() == 39000
assert result["Count"].iloc[0] == 4342

## Groupby

Let's now turn to using groupby from lecture 4.

**Note:** This [slide](https://docs.google.com/presentation/d/1FC-cs5MTGSkDzI_7R_ZENgwoHQ4aVamxFOpJuWT0fo0/edit#slide=id.g477ed0f02e_0_390) provides a visual picture of how `groupby.agg` works if you'd like a reference.

### Question 6: Elections

**Review**: Let's start by reading in the election dataset from the pandas lectures.

In [376]:
# run this cell
elections = pd.read_csv("data/elections.csv")
elections.head(5)

Year          Candidate                  Party  Popular vote Result  \
0  1824     Andrew Jackson  Democratic-Republican        151271   loss   
1  1824  John Quincy Adams  Democratic-Republican        113142    win   
2  1828     Andrew Jackson             Democratic        642806    win   
3  1828  John Quincy Adams    National Republican        500897   loss   
4  1832     Andrew Jackson             Democratic        702735    win   

           %  
0  57.210122  
1  42.789878  
2  56.203927  
3  43.796073  
4  54.574789

As we saw, we can groupby a specific column, e.g. "Party". It turns out that using some syntax we didn't cover in lecture, we can print out the subframes that result. This isn't something you'll do for any practical purpose. However, it may help you get an understanding of what groupby is actually doing.

An example is given below for elections since 1980.

In [377]:
# run this cell
for n, g in elections.query("Year >= 1980").groupby("Party"):
    print(f"Name: {n}") # by the way this is an "f string", a relatively new and great feature of Python
    display(g)

Name: Citizens


Year       Candidate     Party  Popular vote Result         %
127  1980  Barry Commoner  Citizens        233052   loss  0.270182

Name: Constitution


Year         Candidate         Party  Popular vote Result         %
160  2004  Michael Peroutka  Constitution        143630   loss  0.117542
164  2008     Chuck Baldwin  Constitution        199750   loss  0.152398
172  2016    Darrell Castle  Constitution        203091   loss  0.149640

Name: Democratic


Year        Candidate       Party  Popular vote Result          %
129  1980     Jimmy Carter  Democratic      35480115   loss  41.132848
134  1984   Walter Mondale  Democratic      37577352   loss  40.729429
137  1988  Michael Dukakis  Democratic      41809074   loss  45.770691
140  1992     Bill Clinton  Democratic      44909806    win  43.118485
144  1996     Bill Clinton  Democratic      47400125    win  49.296938
151  2000          Al Gore  Democratic      50999897   loss  48.491813
158  2004       John Kerry  Democratic      59028444   loss  48.306775
162  2008     Barack Obama  Democratic      69498516    win  53.023510
168  2012     Barack Obama  Democratic      65915795    win  51.258484
176  2016  Hillary Clinton  Democratic      65853514   loss  48.521539
178  2020     Joseph Biden  Democratic      81268924    win  51.311515

Name: Green


Year         Candidate  Party  Popular vote Result         %
149  1996       Ralph Nader  Green        685297   loss  0.712721
155  2000       Ralph Nader  Green       2882955   loss  2.741176
156  2004        David Cobb  Green        119859   loss  0.098088
165  2008  Cynthia McKinney  Green        161797   loss  0.123442
170  2012        Jill Stein  Green        469627   loss  0.365199
177  2016        Jill Stein  Green       1457226   loss  1.073699
181  2020    Howard Hawkins  Green        405035   loss  0.255731

Name: Independent


Year         Candidate        Party  Popular vote Result          %
130  1980  John B. Anderson  Independent       5719850   loss   6.631143
143  1992        Ross Perot  Independent      19743821   loss  18.956298
161  2004       Ralph Nader  Independent        465151   loss   0.380663
167  2008       Ralph Nader  Independent        739034   loss   0.563842
174  2016     Evan McMullin  Independent        732273   loss   0.539546

Name: Libertarian


Year         Candidate        Party  Popular vote Result         %
128  1980          Ed Clark  Libertarian        921128   loss  1.067883
132  1984    David Bergland  Libertarian        228111   loss  0.247245
138  1988          Ron Paul  Libertarian        431750   loss  0.472660
139  1992      Andre Marrou  Libertarian        290087   loss  0.278516
146  1996      Harry Browne  Libertarian        485759   loss  0.505198
153  2000      Harry Browne  Libertarian        384431   loss  0.365525
159  2004  Michael Badnarik  Libertarian        397265   loss  0.325108
163  2008          Bob Barr  Libertarian        523715   loss  0.399565
169  2012      Gary Johnson  Libertarian       1275971   loss  0.992241
175  2016      Gary Johnson  Libertarian       4489235   loss  3.307714
180  2020      Jo Jorgensen  Libertarian       1865724   loss  1.177979

Name: Natural Law


Year     Candidate        Party  Popular vote Result         %
148  1996  John Hagelin  Natural Law        113670   loss  0.118219

Name: New Alliance


Year      Candidate         Party  Popular vote Result         %
136  1988  Lenora Fulani  New Alliance        217221   loss  0.237804

Name: Populist


Year Candidate     Party  Popular vote Result         %
141  1992  Bo Gritz  Populist        106152   loss  0.101918

Name: Reform


Year     Candidate   Party  Popular vote Result         %
150  1996    Ross Perot  Reform       8085294   loss  8.408844
154  2000  Pat Buchanan  Reform        448895   loss  0.426819

Name: Republican


Year          Candidate       Party  Popular vote Result          %
131  1980      Ronald Reagan  Republican      43903230    win  50.897944
133  1984      Ronald Reagan  Republican      54455472    win  59.023326
135  1988  George H. W. Bush  Republican      48886597    win  53.518845
142  1992  George H. W. Bush  Republican      39104550   loss  37.544784
145  1996           Bob Dole  Republican      39197469   loss  40.766036
152  2000     George W. Bush  Republican      50456002    win  47.974666
157  2004     George W. Bush  Republican      62040610    win  50.771824
166  2008        John McCain  Republican      59948323   loss  45.737243
171  2012        Mitt Romney  Republican      60933504   loss  47.384076
173  2016       Donald Trump  Republican      62984828    win  46.407862
179  2020       Donald Trump  Republican      74216154   loss  46.858542

Name: Taxpayers


Year        Candidate      Party  Popular vote Result         %
147  1996  Howard Phillips  Taxpayers        184656   loss  0.192045

Recall that once we've formed groups, we can aggregate each sub-dataframe (a.k.a. group) into a single row using an aggregation function. For example, if we use `.agg(np.mean)` on the groups above, we get back a single DataFrame where each group has been replaced by a single row. In each column for that aggregate row, the value that appears is the average of all values in that group.

For columns which are non-numeric, e.g. "Result", the column is dropped because we cannot compute the mean of the Result.

In [378]:
elections.query("Year >= 1980").groupby("Party").agg(np.mean)

Year  Popular vote          %
Party                                             
Citizens      1980.000000  2.330520e+05   0.270182
Constitution  2009.333333  1.821570e+05   0.139860
Democratic    2000.000000  5.452196e+07  47.360184
Green         2008.000000  8.831137e+05   0.767151
Independent   2000.000000  5.480026e+06   5.414298
Libertarian   2000.000000  1.026652e+06   0.830876
Natural Law   1996.000000  1.136700e+05   0.118219
New Alliance  1988.000000  2.172210e+05   0.237804
Populist      1992.000000  1.061520e+05   0.101918
Reform        1998.000000  4.267094e+06   4.417831
Republican    2000.000000  5.419334e+07  47.898650
Taxpayers     1996.000000  1.846560e+05   0.192045

Equivalently we can use one of the shorthand aggregation functions, e.g. `.mean()`:

In [379]:
elections.query("Year >= 1980").groupby("Party").mean()

Year  Popular vote          %
Party                                             
Citizens      1980.000000  2.330520e+05   0.270182
Constitution  2009.333333  1.821570e+05   0.139860
Democratic    2000.000000  5.452196e+07  47.360184
Green         2008.000000  8.831137e+05   0.767151
Independent   2000.000000  5.480026e+06   5.414298
Libertarian   2000.000000  1.026652e+06   0.830876
Natural Law   1996.000000  1.136700e+05   0.118219
New Alliance  1988.000000  2.172210e+05   0.237804
Populist      1992.000000  1.061520e+05   0.101918
Reform        1998.000000  4.267094e+06   4.417831
Republican    2000.000000  5.419334e+07  47.898650
Taxpayers     1996.000000  1.846560e+05   0.192045

Note that the index of the dataframe returned by an `groupby.agg` call is no longer a set of numeric indices from 0 to N-1. Instead, we see that the index for the example above is now the `Party`. If we want to restore our DataFrame so that `Party` is a column rather than the index, we can use `reset_index`.

In [380]:
elections.query("Year >= 1980").groupby("Party").mean().reset_index()

Party         Year  Popular vote          %
0       Citizens  1980.000000  2.330520e+05   0.270182
1   Constitution  2009.333333  1.821570e+05   0.139860
2     Democratic  2000.000000  5.452196e+07  47.360184
3          Green  2008.000000  8.831137e+05   0.767151
4    Independent  2000.000000  5.480026e+06   5.414298
5    Libertarian  2000.000000  1.026652e+06   0.830876
6    Natural Law  1996.000000  1.136700e+05   0.118219
7   New Alliance  1988.000000  2.172210e+05   0.237804
8       Populist  1992.000000  1.061520e+05   0.101918
9         Reform  1998.000000  4.267094e+06   4.417831
10    Republican  2000.000000  5.419334e+07  47.898650
11     Taxpayers  1996.000000  1.846560e+05   0.192045

**IMPORTANT NOTE:** You should NEVER NEVER solve problems like the one above using loops or list comprehensions. This is slow and also misses the entire point of this part of DS100.

Before we continue, we'll print out the election dataset again for your convenience.

In [381]:
elections.head(5)

Year          Candidate                  Party  Popular vote Result  \
0  1824     Andrew Jackson  Democratic-Republican        151271   loss   
1  1824  John Quincy Adams  Democratic-Republican        113142    win   
2  1828     Andrew Jackson             Democratic        642806    win   
3  1828  John Quincy Adams    National Republican        500897   loss   
4  1832     Andrew Jackson             Democratic        702735    win   

           %  
0  57.210122  
1  42.789878  
2  56.203927  
3  43.796073  
4  54.574789

### Question 6a
Using `groupby.agg` or one of the shorthand methods (`groupby.min`, `groupby.first`, etc.), create a Series `best_result_percentage_only` that returns a Series showing the entire best result for every party, sorted in decreasing order. Your Series should include only parties which have earned at least 10% of the vote in some election. Your result should look like this:

<code>
Party
Democratic               61.344703
Republican               60.907806
Democratic-Republican    57.210122
National Union           54.951512
Whig                     53.051213
Liberal Republican       44.071406
National Republican      43.796073
Northern Democratic      29.522311
Progressive              27.457433
American                 21.554001
Independent              18.956298
Southern Democratic      18.138998
American Independent     13.571218
Constitutional Union     12.639283
Free Soil                10.138474
Name: %, dtype: float64
</code>
<br/>

A list of named `groupby.agg` shorthand methods is [here](https://pandas.pydata.org/pandas-docs/stable/user_guide/groupby.html#aggregation) (you'll have to scroll down about one page).

<!--
BEGIN QUESTION
name: q6a
-->

In [382]:
best_result_percentage_only = elections[elections['%'] > 10].groupby('Party').agg(max)['%'].sort_values(ascending=False)
best_result_percentage_only = elections[elections['%'] > 10].groupby('Party').agg(max).sort_values('%', ascending=False)['%']
# put your code above this line
best_result_percentage_only

Party
Democratic               61.344703
Republican               60.907806
Democratic-Republican    57.210122
National Union           54.951512
Whig                     53.051213
Liberal Republican       44.071406
National Republican      43.796073
Northern Democratic      29.522311
Progressive              27.457433
American                 21.554001
Independent              18.956298
Southern Democratic      18.138998
American Independent     13.571218
Constitutional Union     12.639283
Free Soil                10.138474
Name: %, dtype: float64

In [383]:
assert len(best_result_percentage_only) == 15
assert best_result_percentage_only["Independent"].sum() == 18.95629754
assert best_result_percentage_only.iloc[0] == 61.34470329

### Question 6c

Our DataFrame contains a number of parties which have never had a successful presidential run. For example, the 2020 elections included candiates from the Libertarian and Green parties, neither of which have elected a president.

In [384]:
# just run this cell
elections.tail(5)

Year       Candidate        Party  Popular vote Result          %
177  2016      Jill Stein        Green       1457226   loss   1.073699
178  2020    Joseph Biden   Democratic      81268924    win  51.311515
179  2020    Donald Trump   Republican      74216154   loss  46.858542
180  2020    Jo Jorgensen  Libertarian       1865724   loss   1.177979
181  2020  Howard Hawkins        Green        405035   loss   0.255731

Suppose we were conducting an analysis trying to focus our attention on parties that had elected a president.

The most natural approach is to use `groupby.filter`. This is an incredibly powerful but subtle tool for filtering data.

As a reminder of how filter works, see [this slide](https://docs.google.com/presentation/d/1FC-cs5MTGSkDzI_7R_ZENgwoHQ4aVamxFOpJuWT0fo0/edit#slide=id.g5ff184b7f5_0_507).
The code below accomplishes the task at hand. It does this by creating a function that returns True if and only if a sub-dataframe (a.k.a. group) contains at least one winner. This function in turn uses the [Pandas function "any"](https://pandas.pydata.org/docs/reference/api/pandas.Series.any.html).

In [385]:
# just run this cell
def at_least_one_candidate_in_the_frame_has_won(frame):
    """Returns df with rows only kept for parties that have
    won at least one election
    """
    return (frame["Result"] == 'win').any()

winners_only = (
    elections
        .groupby("Party")
        .filter(at_least_one_candidate_in_the_frame_has_won)
)
winners_only.tail(5)

Year        Candidate       Party  Popular vote Result          %
171  2012      Mitt Romney  Republican      60933504   loss  47.384076
173  2016     Donald Trump  Republican      62984828    win  46.407862
176  2016  Hillary Clinton  Democratic      65853514   loss  48.521539
178  2020     Joseph Biden  Democratic      81268924    win  51.311515
179  2020     Donald Trump  Republican      74216154   loss  46.858542

Alternately we could have used a `lambda` function instead of explicitly defining a named function using `def`.

In [386]:
# just run this cell (alternative)
winners_only = (
    elections
        .groupby("Party")
        .filter(lambda x : (x["Result"] == "win").any())
)
winners_only.tail(5)

Year        Candidate       Party  Popular vote Result          %
171  2012      Mitt Romney  Republican      60933504   loss  47.384076
173  2016     Donald Trump  Republican      62984828    win  46.407862
176  2016  Hillary Clinton  Democratic      65853514   loss  48.521539
178  2020     Joseph Biden  Democratic      81268924    win  51.311515
179  2020     Donald Trump  Republican      74216154   loss  46.858542

For your exercise, you'll do a less restrictive filtering of the elections data.

**Exercise**: Using `filter`, create a DataFrame `major_party_results_since_1988` that includes all election results starting in 1988, but only show a row if the Party it belongs to has earned at least 1% of the popular vote in ANY election since 1988.

For example, in 1988, you should not include the `New Alliance` candidate, since this party has not earned 1% of the vote since 1988. However, you should include the `Libertarian` candidate from 1988 despite only having 0.47 percent of the vote in 1988, because in 2016 and 2020, the Libertarian candidates Gary Johnson and Jo Jorgensen exceeded 1% of the vote.

For example, the first three rows of the table you generate should look like:

|     |   Year | Candidate         | Party       |   Popular vote | Result   |         % |
|----:|-------:|:------------------|:------------|---------------:|:---------|----------:|
| 135 |   1988 | George H. W. Bush | Republican  |       48886597 | win      | 53.5188   |
| 137 |   1988 | Michael Dukakis   | Democratic  |       41809074 | loss     | 45.7707   |
| 138 |   1988 | Ron Paul          | Libertarian |         431750 | loss     |  0.47266  |

<!--
BEGIN QUESTION
name: q6c
-->

In [387]:
major_party_results_since_1988 = elections[elections['Year'] >= 1988].groupby('Party').filter(lambda group: group['%'].max() > 1)
major_party_results_since_1988.head()

Year          Candidate        Party  Popular vote Result          %
135  1988  George H. W. Bush   Republican      48886597    win  53.518845
137  1988    Michael Dukakis   Democratic      41809074   loss  45.770691
138  1988           Ron Paul  Libertarian        431750   loss   0.472660
139  1992       Andre Marrou  Libertarian        290087   loss   0.278516
140  1992       Bill Clinton   Democratic      44909806    win  43.118485

In [388]:
assert major_party_results_since_1988.shape == (40, 6)
assert major_party_results_since_1988["%"].min() == 0.098088334
assert major_party_results_since_1988["Candidate"].iloc[0] == 'George H. W. Bush'

### Question 7

Pandas provides special purpose functions for working with specific common data types such as strings and dates. For example, the code below provides the length of every Candidate's name from our elections dataset.

In [389]:
elections["Candidate"].str.len()

0      14
1      17
2      14
3      17
4      14
       ..
177    10
178    12
179    12
180    12
181    14
Name: Candidate, Length: 182, dtype: int64

**Exercise**: Using `.str.split`. Create a new DataFrame called `elections_with_first_name` with a new column `First Name` that is equal to the Candidate's first name.

See the Pandas `str` [documentation](https://pandas.pydata.org/docs/reference/api/pandas.Series.str.split.html) for documentation on using `str.split`.

Hint: Use `[0]` somewhere in your code.

<!--
BEGIN QUESTION
name: q7
-->

In [391]:
elections_with_first_name = elections.copy()
# your code here
elections_with_first_name['First Name'] = elections_with_first_name['Candidate'].str.split().str[0]
# end your code
elections_with_first_name

Year          Candidate                  Party  Popular vote Result  \
0    1824     Andrew Jackson  Democratic-Republican        151271   loss   
1    1824  John Quincy Adams  Democratic-Republican        113142    win   
2    1828     Andrew Jackson             Democratic        642806    win   
3    1828  John Quincy Adams    National Republican        500897   loss   
4    1832     Andrew Jackson             Democratic        702735    win   
..    ...                ...                    ...           ...    ...   
177  2016         Jill Stein                  Green       1457226   loss   
178  2020       Joseph Biden             Democratic      81268924    win   
179  2020       Donald Trump             Republican      74216154   loss   
180  2020       Jo Jorgensen            Libertarian       1865724   loss   
181  2020     Howard Hawkins                  Green        405035   loss   

             % First Name  
0    57.210122     Andrew  
1    42.789878       John  
2    56.203927     Andrew  
3    43.796073       John  
4    54.574789     Andrew  
..         ...        ...  
177   1.073699       Jill  
178  51.311515     Joseph  
179  46.858542     Donald  
180   1.177979         Jo  
181   0.255731     Howard  

[182 rows x 7 columns]

In [392]:
assert elections_with_first_name.shape == (182, 7)
assert elections_with_first_name[elections_with_first_name["Candidate"] == "Andrew Jackson"].iloc[0]["First Name"] == 'Andrew'
assert elections_with_first_name[elections_with_first_name["Candidate"] == "Jo Jorgensen"].iloc[0]["First Name"] == 'Jo'

### Question 8

The code below creates a table with the frequency of all names from 2020.

In [393]:
# just run this cell
baby_names_2020 = (
    baby_names.query('Year == 2020')
              .groupby("Name")
              .sum()[["Count"]]
              .reset_index()
)
baby_names_2020

Name  Count
0       Aaden     15
1     Aadhira      6
2     Aadhvik      5
3      Aadhya    186
4        Aadi     14
...       ...    ...
8697   Zymere      6
8698    Zymir     74
8699     Zyon    130
8700     Zyra     33
8701    Zyrah      5

[8702 rows x 2 columns]

**Exercise**: Using the `pd.merge` function, see the [documentation](https://pandas.pydata.org/docs/reference/api/pandas.merge.html), combine the `baby_names_2020` table with the `elections_with_first_name` table you created earlier to form `presidential_candidates_and_name_popularity`.

<!--
BEGIN QUESTION
name: q8
-->

In [394]:
presidential_candidates_and_name_popularity = baby_names_2020.merge(elections_with_first_name, left_on = 'Name', right_on = 'First Name')
presidential_candidates_and_name_popularity

Name  Count  Year          Candidate           Party  Popular vote  \
0      Aaron   5032  1920   Aaron S. Watkins     Prohibition        188787   
1    Abraham   1851  1860    Abraham Lincoln      Republican       1855993   
2    Abraham   1851  1864    Abraham Lincoln  National Union       2211317   
3      Allan    222  1916    Allan L. Benson       Socialist        590524   
4      Alton     35  1904    Alton B. Parker      Democratic       5083880   
..       ...    ...   ...                ...             ...           ...   
148  William  12541  1932  William Z. Foster       Communist        103307   
149  William  12541  1936      William Lemke           Union        892378   
150  Woodrow     23  1912     Woodrow Wilson      Democratic       6296284   
151  Woodrow     23  1916     Woodrow Wilson      Democratic       9126868   
152  Zachary   2684  1848     Zachary Taylor            Whig       1360235   

    Result          % First Name  
0     loss   0.708351      Aaron  
1      win  39.699408    Abraham  
2      win  54.951512    Abraham  
3     loss   3.194193      Allan  
4     loss  37.685116      Alton  
..     ...        ...        ...  
148   loss   0.261069    William  
149   loss   1.960733    William  
150    win  41.933422    Woodrow  
151    win  49.367987    Woodrow  
152    win  47.309296    Zachary  

[153 rows x 9 columns]

In [395]:
assert presidential_candidates_and_name_popularity.shape == (153, 9)
assert presidential_candidates_and_name_popularity[presidential_candidates_and_name_popularity["Candidate"] == "Andrew Jackson"].iloc[0]["Name"] == 'Andrew'
assert presidential_candidates_and_name_popularity[presidential_candidates_and_name_popularity["Candidate"] == "Jo Jorgensen"].iloc[0]["Count"] == 6

---

# Well done!